### PR TITLE
[opt] : when composite primary key has a column that is auto-incr, and this column does not specify a value, no dup check is needed

### PR DESCRIPTION
**Pull request by jensenojs**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #15809

## What this PR does / why we need it:

when composite primary key has a column that is auto-incr, and this column does not specify a value, no dup check is needed

### CODE REVIEW

**Review by fengttt — CHANGES_REQUESTED** [review]
For this PR, I have a concern on correctness.

If we disable the dup check, and someone insert a row with a value, later, auto_incr reuse that value, we will have a data corruption.

**Comment by jensenojs** [comment]
> For this PR, I have a concern on correctness.
> 
> If we disable the dup check, and someone insert a row with a value, later, auto_incr reuse that value, we will have a data corruption.

if we allow user insert an id value, we invalidate all id caches and start from a value that is bigger than this id. We have already done this and it was confirmed by @zhangxu19830126

**Comment by jensenojs** [comment]
We can supplement relevant bvt test cases

**Comment by fengttt** [comment]
> > For this PR, I have a concern on correctness.
> > If we disable the dup check, and someone insert a row with a value, later, auto_incr reuse that value, we will have a data corruption.
> 
> if we allow user insert an id value, we invalidate all id caches and start from a value that is bigger than this id. We have already done this and it was confirmed by @zhangxu19830126

Now,
1. add test in bvt that proves you are correct.
2. This is a high risk, high impact change.   So I ask you please to add a table level config, so that we can enable/disable.    I would like this config to be OFF by default, but you may insist this to be ON by default.

**Comment by jensenojs** [comment]
我理解这里可能有风险的根本原因是, 
- 针对自增类型列的插入
  - 如果允许在插入的时候指定值
    - `incr service`能不能让对应的`column cache`失效, 并从插入的值中最大的值开始自增

上面这个事情有一个足够强的测试去覆盖这个场景, 当这里有bug的时候, 那就需要去重去兜底. 

这个机制在mo中具体实现的位置大概在 : https://github.com/matrixorigin/matrixone/blob/9e1107c34ae4606649c4465f907220f0da4618ee/pkg/incrservice/column_cache.go#L548

`incr service` 会维护插入数据中有值的最大值, 然后拿去更新. 但在`column_cache.test`中好像确实没有关于这方面的测试, 即并发地调用`insertAutoValues`, 观察`lastInsertValue`是否正确 —— 即大于`insert`时指定/没指定的所有值

所以在正式开始配置前, 我补充了相关的bvt和ut测试, 其中`ut`可以做到高并发的检测, bvt只有两个session, 后面可能需要继续补充. 

在实现

> > > For this PR, I have a concern on correctness.
> > > If we disable the dup check, and someone insert a row with a value, later, auto_incr reuse that value, we will have a data corruption.
> > 
> > 
> > if we allow user insert an id value, we invalidate all id caches and start from a value that is bigger than this id. We have already done this and it was confirmed by @zhangxu19830126
> 
> Now,
> 
> 1. add test in bvt that proves you are correct.
> 2. This is a high risk, high impact change.   So I ask you please to add a table level config, so that we can enable/disable.    I would like this config to be OFF by default, but you may insist this to be ON by default.

**Comment by fengttt** [comment]
> 在实现配置之前, 想再向田博士 @fengttt 请教和确认一下
> 
> * 如果是`table level config`的话, 假设有N个表, 为了测不是默认配置的场景, 是不是需要手动配置`N`次之后才能开始测试? 不确定我理解错了没有
>   
>   * 如果是这样的话好像会有点啰嗦, `session level`的配置会不会简单一些?

If you can, this should be a table level config.   We accept the manual work if we need to configure it on several tables during test.   

This should not a session level config.   Cannot depend on user to set this one correctly.

File: pkg/frontend/authenticate.go [hunk 1]
```diff
@@ -9275,6 +9309,46 @@ func GetVersionCompatibility(ctx context.Context, ses *Session, dbName string) (
 	return resultConfig, err
 }
 
+func GetUniqueCheckOnAutoIncr(ctx context.Context, ses FeSession, dbName string) (ret string, err error) {
```
Comment (fengttt) [inline pkg/frontend/authenticate.go:9312]: Why do you need this?  Do we need to write such a function for EVERY config we add?   Feels wrong.

File: test/distributed/cases/dml/insert/insert_auto.sql
[the remark this answers is not in the document]
Comment (fengttt) [reply test/distributed/cases/dml/insert/insert_auto.sql:1]: You are inserting a value 10, which will not cause a conflict.   So use 3.  
And insert some again in step 4.

**Review by fengttt — CHANGES_REQUESTED** [review]

File: test/distributed/cases/dml/insert/insert_auto.sql
[the remark this answers is not in the document]
Comment (jensenojs) [reply test/distributed/cases/dml/insert/insert_auto.sql:1]:
```sql
create table t0(a int auto_increment, b int, primary key(a));
insert into t0(b) values (1), (2);
```

下一个`insert`
- 如果没有指定值
- 如果指定的值是3

对于上两种情况而言都是没有区别的. 
```sql
mysql> create table t0(a int auto_increment, b int, primary key(a));
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t0(b) values (1), (2);
Query OK, 2 rows affected (0.01 sec)

mysql> insert into t0(a, b) values (3, 3);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t0(b) values (1), (2);
Query OK, 2 rows affected (0.01 sec)

mysql> select * from t0;
+------+------+
| a    | b    |
+------+------+
|    1 |    1 |
|    2 |    2 |
|    3 |    3 |
|    4 |    1 |
|    5 |    2 |
+------+------+
5 rows in set (0.00 sec)

mysql> 
```

所以我才考虑插入一个比3大的数值, 然后在插入成功之后, 检验在此之后的自增是否是基于手动插入的更大的数的基础上开始的

File: pkg/frontend/authenticate.go (hunk 1 above)
Comment (jensenojs) [reply pkg/frontend/authenticate.go:9312]: 确实, 很乱... 但向杨钊哥请教了一下, 目前的实现就是这个样子的

DB level级别的config实际上还都是在`authenticate.go`中手动set进去的...

https://github.com/matrixorigin/matrixone/blob/2558e02239416ca4ad2a136a33f2496450318aa7/pkg/frontend/authenticate.go#L9150

上面的代码会在创建一个database的时候, 手动将一些变量给塞进`MysqlCompatbilityMode`中, 用一个sql表去维护所有的变量, 所以每次都需要通过后台SQL去读取这个值有没有更新

我不确定每次执行insert之前调用一个后台SQL有没有问题, 但变量的管理维护挺乱的..
@fengttt

File: test/distributed/cases/dml/insert/insert_auto.sql
[the remark this answers is not in the document]
Comment (fengttt) [reply test/distributed/cases/dml/insert/insert_auto.sql:1]: I think insert (3, 3) is the most interesting test.    Repeat this with all different configurations.   Including check/not check/fail.

**Review by fengttt — CHANGES_REQUESTED** [review]
Test has no coverage on the config.   Please add.

**Comment by jensenojs** [comment]
还有一点需要补充的是, 目前`auto incr unique key`是没有这个优化的, 也就是说`unique key`无论如何都会进行去重, 已经建立了相关的issue 进行跟踪, 后续close对应issue的时候会补充相关的测试用例
- #15958

**Comment by jensenojs** [comment]
基于`ff69679fd7`, 用`sysbench_insert`在, table_size=100000, threads=100, time=60, auto_inc=on, create_secondary=on的情况下, main的qps为7039.12 per sec, 本pr的qps为5038.30 per sec (只对比了一次), **性能为原来的71%**

慢的根本原因就是在`insert`的时候需要通过后台SQL获取DB level config, 不用后台SQL多CN会有问题, 而且会破坏目前`System variable`相关的设计, 用后台SQL就会慢. @fengttt 

在解决性能问题之前不能合并

main : 
Pasted Graphic 28

this pr : 
Pasted Graphic 29

**Comment by jensenojs** [comment]
# 性能对比 : 

sysbench-insert 在深圳的台式机上跑了三次, 平均值上看和main已经没有明显差异, 明天补充一个更完整的性能对比..

```shell
sysbench --mysql-host=127.0.0.1 --mysql-port=6001 --mysql-user=dump --mysql-password=111 oltp_common.lua --mysql-db=sysbench_db --tables=10 --table_size=0 --threads=1 --time=30 --report-interval=10 --create_secondary=on --auto_inc=on prepare

sysbench --mysql-host=127.0.0.1 --mysql-port=6001 --mysql-user=dump --mysql-password=111 --db-ps-mode=disable oltp_insert.lua --mysql-db=sysbench_db --tables=10 --table_size=100000 --threads=100 --time=60 --report-interval=10 --create_secondary=on --auto_inc=off --skip_trx=on --range_selects=off --point_selects=1 run
```

# 配置行为描述

> Just that before the final review, update your PR comments.    Clearly state the config, what are the values, what is the default, how do you expect customer to use it, the point the expected use case to a test case.    Let's finish this.

TODO,  写完之后再请大家review

# 遗留工作

这个pr给前端添加了一个获取config的接口, 和`resloveVariable`那一套的体系是分开的, 第一个`DB level config`的配置编写是硬hack的, 所以现在的实现其实也很不优雅, 需要重构, 不然后面table level config更不可能实现, 下面的issue 待分配.

- https://github.com/matrixorigin/matrixone/issues/15809

**Comment by jensenojs** [comment]
现在`restore`相关的bvt并不稳定, 而且这个pr的内容也太大了, 我会把它拆成两个部分的内容去提交. 

- Database level config
- unique_check_on_autoincr 的配置

**Pull request closed without merging**